### PR TITLE
[GUI][SKIN] ContextWindow propagation for nested skin variables

### LINKED ISSUE #15348: Skinning - ListItem infolabels in nested variables do not populate when a dialog is visible.
### Describe the bug
If a listitem infolabel is in a variable nested in another variable, then the info will not show when a dialog such as contextmenu is on screen.

## Expected Behavior
The info will show even if a dialog is on screen.

## Actual Behavior
The info does not show.

### To Reproduce
1. Add a label control in MyVideoNav which references a variable:
```
<control type="label">
<label>$VAR[var_foo]</label>
</control>
```

In an includes file add the variable:
```
<variable name="var_foo">
<value>$VAR[var_bar]</value>
</variable>

<variable name="var_bar">
<value>$INFO[ListItem.Label]</value>
</variable>
```

The label will show when the library is visible. Open the context menu and it disappears. However, if you reference $VAR[var_bar] directly from the label control instead, then the info will stay.

## Additional context 
Nested variables help cut down on code sizes considerably. For instance, it is necessary to nest a variable that converts audio channel numbers into the correct speaker configuration (e.g. 2 = 2.0, 3 = 2.1, 4 = 3.1 etc.). Without a nested variable, complex labels need a condition for every single possible configuration.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [x] Windows
 - [ ] Windows UWP

The bug happens on Linux (Fedora), Windows 10, and Libreelec. (i.e. not platform dependent).
 - Kodi version: RC5.2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by enen92**
## Description
Kodi supports nested skin variables however it doesn't propagate the window context when evaluating them. This might lead to the label to stop displaying once some other dialog pops up over the dialog where the label should be evaluated.
@ronie, @jurialmunkey. @HitcherUK (and other skinners) would be nice if you could runtime test this to see if the issue is fixed and nothing else pops up (it didn't in my tests).

## Motivation and context
Fix https://github.com/xbmc/xbmc/issues/15348

## How has this been tested?
Runtime tested with the procedure outlined in the bug report

## What is the effect on users?
Simple skin fix, some infolabels belonging to nested skin variables won't be discarded anymore

### CODE REVIEW

**Comment by jurialmunkey** [comment]
WIN64 testbuild 
http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20220423-a96cdcd9-nestedskin-x64.exe

Infolabels in `<value> </value>` for nested variables work now. However, the condition tags `<value condition="">` still don't work.

Test XML:

Home.xml
```
<control type="label">
<label>$VAR[var_1]</label>
</control>
```

Includes.xml
```
<variable name="var_1">
    <value>$VAR[var_2]</value>
</variable>
<variable name="var_2">
    <value condition="!String.IsEmpty(ListItem.Label)">var_2 $INFO[ListItem.Label] $VAR[var_3]</value>
    <value>var_2 FAIL $VAR[var_3]</value>
</variable>
<variable name="var_3">
    <value condition="!String.IsEmpty(ListItem.Label)">var_3 $INFO[ListItem.Label]</value>
    <value>var_3 FAIL $INFO[ListItem.Label]</value>
</variable>
```

Tested with DialogContextMenu, DialogButtonMenu, and DialogConfirm. Variable conditions fail on ContextMenu and Confirm but work on ButtonMenu.

**Review by jurialmunkey — COMMENTED** [review]
See comment - https://github.com/xbmc/xbmc/pull/21303#issuecomment-1107688219

**Comment by enen92** [comment]
Hum thanks, I can confirm conditions don't evaluate correctly with your sample. Will check if I can figure it out

**Comment by enen92** [comment]
Last commit ended up blowing the diff since the *confusing* default value for `contextWindow` was removed and moved to all the callees. Important thing here for the review is the following:

- Skin variables and info expressions are constructed with `DEFAULT_CONTEXT` by default (just like in the past) if they are not bound to a a specific context/window. Think of `DEFAULT_CONTEXT` as something generic that indicates those constructs might have different values depending on the context they are called.
- When evaluating those items if they have the `DEFAULT_CONTEXT` the callee context is used instead

**Comment by enen92** [comment]
@jurialmunkey please retest if possible and report back. The second commit adds the same approach for the info expressions as well

File: xbmc/interfaces/builtins/PVRBuiltins.cpp [hunk 1]
```diff
@@ -66,10 +67,12 @@ static int SeekPercentage(const std::vector<std::string>& params)
       CGUIInfoManager& infoMgr = CServiceBroker::GetGUI()->GetInfoManager();
 
       int iTimeshiftProgressDuration = 0;
-      infoMgr.GetInt(iTimeshiftProgressDuration, PVR_TIMESHIFT_PROGRESS_DURATION);
+      infoMgr.GetInt(iTimeshiftProgressDuration, PVR_TIMESHIFT_PROGRESS_DURATION,
```
Comment (enen92) [inline xbmc/interfaces/builtins/PVRBuiltins.cpp:70]: @ksooo @phunkyfish it'd be nice if calling GUI/InfoManager could be somehow avoided here (this info is probably available in the PVRManager itself. `EpgGridControl()` also seem to rely on GUI maybe the main logic can be also moved to `PVRGUIActions` without much effort?

**Comment by jurialmunkey** [comment]
> @jurialmunkey please retest if possible and report back. The second commit adds the same approach for the info expressions as well

@enen92 - LGTM. All my tests are working now.

I also tried a few more complex position and property references just in case like `Container(id)` `ListItemAbsolute(position)` `ListItem.UniqueID(idname)` `ListItem.Property(property)` and all appears to work as expected.

**Comment by enen92** [comment]
Thanks for testing jurialmunkey. I will wait for some review now.

File: xbmc/guilib/guiinfo/GUIInfoBool.cpp [hunk 2]
```diff
@@ -30,12 +30,12 @@ void CGUIInfoBool::Parse(const std::string &expression, int context)
   else
   {
     m_info = CServiceBroker::GetGUI()->GetInfoManager().Register(expression, context);
-    Update();
+    Update(context);
   }
 }
 
-void CGUIInfoBool::Update(const CGUIListItem *item /*= NULL*/)
+void CGUIInfoBool::Update(int contextWindow, const CGUIListItem* item /*= NULL*/)
```
Comment (howie-f) [inline xbmc/guilib/guiinfo/GUIInfoBool.cpp:37]: nullptr

File: xbmc/guilib/guiinfo/GUIInfoBool.h [hunk 3]
```diff
@@ -34,7 +34,7 @@ class CGUIInfoBool
 
   operator bool() const { return m_value; }
 
-  void Update(const CGUIListItem *item = NULL);
+  void Update(int contextWindow, const CGUIListItem* item = NULL);
```
Comment (howie-f) [inline xbmc/guilib/guiinfo/GUIInfoBool.h:37]: nullptr

File: xbmc/interfaces/info/InfoBool.h [hunk 4]
```diff
@@ -29,15 +29,16 @@ class InfoBool
 
   /*! \brief Get the value of this info bool
    This is called to update (if dirty) and fetch the value of the info bool
+   \param contextWindow the context (window id) where this condition is being evaluated
    \param item the item used to evaluate the bool
    */
-  inline bool Get(const CGUIListItem *item = NULL)
+  inline bool Get(int contextWindow, const CGUIListItem* item = NULL)
```
Comment (howie-f) [inline xbmc/interfaces/info/InfoBool.h:35]: nullptr

**Comment by enen92** [comment]
Latest push moves the `DEFAULT_CONTEXT` to the `INFO` namespace (previously in `GUIINFO` which belongs to `GUILIB`) - IMHO they are not the exactly the same thing and was requiring a few includes I didn't like. Also adjusted @howie-f  comments. I consider this finished and won't touch it anymore :)

**Comment by enen92** [comment]
Some review would be nice. This was confirmed working already by one of our skinners.
I plan to merge by the end of next week unless someone objects

**Pull request merged**